Enable scalac linting / fatal warnings and fix all warnings.

jbgi opened the pull request:
We may not want to include/fix all warnings, but there it is, for further discussion.

jbgi commented on core/src/main/scala/scalaz/EitherT.scala:
```diff
@@ -228,7 +228,7 @@ object EitherT extends EitherTInstances {
   }
 
   def eitherTU[FAB, AB, A0, B0](fab: FAB)(
-    implicit u1: Unapply[Functor, FAB]{type A = AB}, u2: Unapply2[Bifunctor, AB]{type A = A0; type B = B0}, l: Leibniz.===[AB, A0 \/ B0])
```
The unused `Unapply2` was there to guide type inference but it is probably no more necessary... ?

jbgi replied:
oups... looks like it is still necessary.

S11001001 commented on core/src/main/scala/scalaz/LazyEitherT.scala:
```diff
@@ -105,7 +105,7 @@ object LazyEitherT extends LazyEitherTInstances {
     u2: Unapply2[Bifunctor, AB]{type A = A0; type B = B0},
     l: Leibniz.===[AB, LazyEither[A0, B0]]
-  ): LazyEitherT[u1.M, A0, B0] = LazyEitherT(l.subst[u1.M](u1(fab)))
+  ): LazyEitherT[u1.M, u2.A, u2.B] = LazyEitherT(l.subst[u1.M](u1(fab)))
```
Ehh.....this is tricky in the face of expected-type inference (the former propagates while the latter doesn't), best leave it as it was.

jbgi replied:
yeah, this was to use the `u2` parameter, otherwise I get an "unused parameter" warning...
But not sure to understand why this could be a problem for `u2.A` / `u2.B` while it apparently is not for `u1.M`?

S11001001 replied:
It is a problem for `u1.M`, too, but we can't fix it due to scala/bug#5075. (This is *not* scala/bug#2712, despite some references therein and surface similarities.)

We don't have that problem for `A0` and `B0`, so we can Do The Right Thing for those.

S11001001 replied:
That's also why 1110a5e94fbee1f0d324fb70722d9a4dcd7aa008 didn't include an `Aux` variant including the hk member; scala/bug#5075 makes it so it never works.

jbgi replied:
@S11001001 used deprecated annotation to avoid the warning.

S11001001 commented on core/src/main/scala/scalaz/Adjunction.scala:
```diff
@@ -88,7 +88,7 @@ sealed abstract class AdjunctionInstances {
   implicit def curryUncurryAdjunction[S]: (S, ?) -| (S => ?) =
     new Adjunction[(S, ?), (S => ?)] {
-      override def leftAdjunct[A, B](a: => A)(f: ((S, A)) => B): S => B = s => f(s, a)
+      override def leftAdjunct[A, B](a: => A)(f: ((S, A)) => B): S => B = s => f((s, a))
```
Huh, could have sworn we had `-Yno-adapted-args` a long time ago.  Oh well 😎

S11001001 commented on core/src/main/scala/scalaz/FingerTree.scala:
```diff
@@ -807,18 +805,6 @@ sealed abstract class FingerTree[V, A](implicit measurer: Reducer[A, V]) {
     )
   }
 
-  private def traverseFinger[F[_], A, B, V2](digit: Finger[V, A])(f: A => F[B])(implicit ms: Reducer[B, V2], F: Applicative[F]): F[Finger[V2, B]] = {
```
Ah, my second favorite reason for using `private` all over the place.  (My first is that incremental compiles are faster 🏎 )

S11001001 commented on core/src/main/scala/scalaz/StateT.scala:
```diff
@@ -268,14 +268,14 @@ private trait StateTMonadState[S, F[_]] extends MonadState[StateT[F, S, ?], S] w
 private trait StateTHoist[S] extends Hoist[λ[(g[_], a) => StateT[g, S, a]]] {
 
-  type StateTF[G[_], S] = {
+  type StateTF[G[_]] = {
```
Nice.

S11001001 commented on core/src/main/scala/scalaz/TracedT.scala:
```diff
@@ -110,7 +110,7 @@ object TracedT extends TracedTInstances {
     U2: Unapply2[Profunctor, AB]{type A = A0; type B = B0},
     L: Leibniz.===[AB, A0 => B0]
-  ): TracedT[U1.M, A0, B0] = TracedT(L.subst[U1.M](U1(wab)))
+  ): TracedT[U1.M, U2.A, U2.B] = TracedT(L.subst[U1.M](U1(wab)))
```
As with other `A0`/`B0` change.

S11001001 commented on project/build.scala:
```diff
@@ -104,12 +104,35 @@ object build {
+    "-Ywarn-dead-code",
+    "-Ywarn-unused-import",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard",
```
Let's turn off `-Ywarn-value-discard` and not have all the extra `()` noise.  That warning's more suited to a heavily imperative codebase.

TomasMikula replied:
Or maybe only turn it on for `core`?

jbgi replied:
@puffnfresh what is your opinion on this option? (as you wrote [NonUnitStatements](http://www.wartremover.org/doc/warts.html#nonunitstatements))

S11001001 commented on core/src/main/scala/scalaz/Validation.scala:
```diff
@@ -34,7 +34,7 @@ import scala.reflect.ClassTag
 
   final class SwitchingValidation[X](s: => X){
-    def <<?:(fail: => X): X =
+    def <<?:(fail: X): X =
```
Whoops, thanks.

S11001001 commented on core/src/main/scala/scalaz/WriterT.scala:
```diff
@@ -284,7 +284,7 @@ trait WriterTFunctions {
     u2: Unapply2[Bifunctor, AB]{type A = A0; type B = B0},
     l: Leibniz.===[AB, (A0, B0)]
-  ): WriterT[u1.M, A0, B0] = WriterT(l.subst[u1.M](u1(fab)))
+  ): WriterT[u1.M, u2.A, u2.B] = WriterT(l.subst[u1.M](u1(fab)))
```
A0/B0 etc etc

xuwei-k commented on project/build.scala:
```diff
@@ -104,12 +104,35 @@ object build {
+    "-Ywarn-unused-import",
+    "-Ywarn-numeric-widen",
+    "-Ywarn-value-discard",
+    "-Xfatal-warnings",
```
remove `-Xfatal-warnings` when generating scaladoc ?

```
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/Validation.scala:6: Could not find any member to link for "a]Validation[E,".
[error] /**
[error] ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/Validation.scala:341: Variable i undefined in comment for method excepting in class Validation
[error]    *   strings map (_.parseInt excepting { case i if i < 0 => new Exception(s"Int must be positive: $i") })
[error]                                                                                                      ^
model contains 44 documentable templates
[error] /home/travis/build/scalaz/scalaz/scalacheck-binding/src/main/scala/scalaz/scalacheck/ScalazArbitrary.scala:9: Could not find any member to link for "scalacheck.Arbitrary".
[error] /**
[error] ^
[error] /home/travis/build/scalaz/scalaz/scalacheck-binding/js/src/main/scala/scalaz/scalacheck/ScalazArbitraryPlatform.scala:3: Could not find any member to link for "scalacheck.Arbitrary".
[error] /**
[error] ^
[error] two errors found
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/Order.scala:38: unbalanced or unclosed heading
[error]   /** @note `Order.fromScalaOrdering(toScalaOrdering).order(x, y)`
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/Free.scala:77: Could not find any member to link for "scalaz.iteratee.Iteratee".
[error]   /** A computation that accepts values of type `A`, eventually resulting in a value of type `B`.
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:768: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:761: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:754: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:747: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:696: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:38: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] /home/travis/build/scalaz/scalaz/core/src/main/scala/scalaz/FingerTree.scala:31: Could not find any member to link for "if".
[error]   /**
[error]   ^
[error] 11 errors found
[error] (scalaz/scalaunidoc:doc) Scaladoc generation failed
[error] (iterateeJS/compile:doc) Scaladoc generation failed
[error] (iterateeJVM/compile:doc) Scaladoc generation failed
[error] (coreJVM/compile:doc) Scaladoc generation failed
[error] (scalacheckBindingJVM/compile:doc) Scaladoc generation failed
[error] (scalacheckBindingJS/compile:doc) Scaladoc generation failed
[error] (coreJS/compile:doc) Scaladoc generation failed
```

https://travis-ci.org/scalaz/scalaz/jobs/226610713#L7169

jbgi commented:
@S11001001 I disabled `-Ywarn-value-discard` for scalacheck-bindings where it causes much noise, and use `val _ = ...` to avoid the extra statement elsewhere. Is that ok?

jbgi commented:
rebased. notable changes since last time: removed unused constraints in:
 - https://github.com/scalaz/scalaz/pull/1373/files#diff-f93de4d0f6d9429dcb912c7dcaff37afL51
 - https://github.com/scalaz/scalaz/pull/1373/files#diff-97816bd0299a1bb451f1415800dc3fd6L16 (/cc @puffnfresh )
